### PR TITLE
Add SignalR Client and scenario

### PR DESCRIPTION
**Pull request by BrennanConroy**
Lets go!

I'd prefer getting this in quickly so we can iterate on it.
It shouldn't cause any issues with our automation and I can spin up a couple test runs on our machines to see what kind of things need work.

### CODE REVIEW

File: src/Benchmarks/signalr.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "Default": {
+    "ClientName": "signalr",
+    "Source": {
+      "Repository": "https://github.com/aspnet/benchmarks.git",
+      "BranchOrCommit": "brecon/signalrClientNew",
```
Comment (BrennanConroy) [inline src/Benchmarks/signalr.json]: Change this before merging

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+                    max = _requestsPerConnection[i];
+                }
+                if (_requestsPerConnection[i] < min)
+                {
+                    min = _requestsPerConnection[i];
+                }
+            }
+            // Review: This could be interesting information, see the gap between most active and least active connection
+            // Ideally they should be within a couple percent of each other, but if they aren't something could be wrong
+            Log($"Least Requests to Connection: {min}");
```
Comment (anurse) [inline src/BenchmarksWorkers/Workers/SignalRWorker.cs]: ~~to~~ **per**

File: src/Benchmarks/Middleware/SignalRMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            return builder.UseSignalR(route =>
+            {
+                route.MapHub<EchoHub>(Scenarios.GetPath(s => s.SignalRBroadcast) + "/default");
+            });
+        }
+    }
+
+    public class EchoHub : Hub
+    {
+        public override Task OnConnectedAsync()
```
Comment (davidfowl) [inline src/Benchmarks/Middleware/SignalRMiddleware.cs]: Remove these overrides?

File: src/Benchmarks/Middleware/SignalRMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            return Task.CompletedTask;
+        }
+
+        public async Task Echo(int duration)
+        {
+            try
+            {
+                var t = new CancellationTokenSource();
+                t.CancelAfter(TimeSpan.FromSeconds(duration));
+                while (!t.IsCancellationRequested)
```
Comment (davidfowl) [inline src/Benchmarks/Middleware/SignalRMiddleware.cs]: Wait on client disconnected as well?

Comment (BrennanConroy) [reply src/Benchmarks/Middleware/SignalRMiddleware.cs]: Ok, I'll keep track of how many connections are on the server and when it hits 0 exit the loop

Comment (davidfowl) [reply src/Benchmarks/Middleware/SignalRMiddleware.cs]: You don't have to. You can just make sure you're not sending after this client is gone by checking this token and the aborted token.

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+            {
+                await WriteJobResultToSqlAsync(serverJob, clientJob, connectionString, tableName, path, session, description, statistics, longRunning, "RequestsPerSecond", statistics.RequestsPerSecond);
+            });
+
+            await RetryOnExceptionAsync(5, async () =>
+            {
+                await WriteJobResultToSqlAsync(serverJob, clientJob, connectionString, tableName, path, session, description, statistics, longRunning, "CPU", statistics.Cpu);
+            });
+
+            await RetryOnExceptionAsync(5, async () =>
```
Comment (davidfowl) [inline src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: I see working set here twice? Copy paste error?

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        private void CreateConnections(TransportType transportType = TransportType.WebSockets)
+        {
+            _connections = new List<HubConnection>(_job.Connections);
+            _requestsPerConnection = new List<int>(_job.Connections);
+            _latencyPerConnection = new List<List<double>>(_job.Connections);
+
+            var hubConnectionBuilder = new HubConnectionBuilder()
+                .WithUrl(_job.ServerBenchmarkUri)
+                .WithMessageHandler(_httpClientHandler)
+                //.WithConsoleLogger(Microsoft.Extensions.Logging.LogLevel.Trace) TODO: Support logging on client for debugging purposes
```
Comment (davidfowl) [inline src/BenchmarksWorkers/Workers/SignalRWorker.cs]: File an issue for this

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+            {
+                _job.State = ClientState.Completed;
+            }
+        }
+
+        public async Task StopAsync()
+        {
+            if (_timer != null)
+            {
+                _timer?.Dispose();
```
Comment (davidfowl) [inline src/BenchmarksWorkers/Workers/SignalRWorker.cs]: We know the timer isn't null here. Do we need a lock?

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+            await Task.WhenAll(tasks);
+
+            _job.State = ClientState.Running;
+            _job.LastDriverCommunicationUtc = DateTime.UtcNow;
+
+            // SendAsync will return as soon as the request has been sent (non-blocking)
+            await _connections[0].SendAsync("Echo", _job.Duration + 1);
+            _workTimer.Start();
+            _timer = new Timer(StopClients, null, TimeSpan.FromSeconds(_job.Duration), Timeout.InfiniteTimeSpan);
```
Comment (davidfowl) [inline src/BenchmarksWorkers/Workers/SignalRWorker.cs]: Wait, we have a timer that runs once?

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs (hunk 5 above)
Comment (BrennanConroy) [reply src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: Yes

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs (hunk 7 above)
Comment (BrennanConroy) [reply src/BenchmarksWorkers/Workers/SignalRWorker.cs]: We should lock

File: src/BenchmarksWorkers/Workers/SignalRWorker.cs (hunk 8 above)
Comment (BrennanConroy) [reply src/BenchmarksWorkers/Workers/SignalRWorker.cs]: A new worker is created for every job, and the timer is to stop the worker when the job is done

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+                        [RuntimeVersion] [nvarchar](50) NOT NULL,
+                        [Scenario] [nvarchar](50) NOT NULL,
+                        [Hardware] [nvarchar](50) NOT NULL,
+                        [HardwareVersion] [nvarchar](128) NOT NULL,
+                        [OperatingSystem] [nvarchar](50) NOT NULL,
+                        [Framework] [nvarchar](50) NOT NULL,
+                        [RuntimeStore] [bit] NOT NULL,
+                        [Scheme] [nvarchar](50) NOT NULL,
+                        [Sources] [nvarchar](50) NULL,
+                        [ConnectionFilter] [nvarchar](50) NULL,
```
Comment (sebastienros) [inline src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: Do you need that?

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+                        [WebHost] [nvarchar](50) NOT NULL,
+                        [Transport] [nvarchar](50) NOT NULL,
+                        [HubProtocol] [nvarchar](50) NOT NULL,
+                        [Connections] [int] NOT NULL,
+                        [Duration] [int] NOT NULL,
+                        [Path] [nvarchar](200) NULL,
+                        [Method] [nvarchar](50) NOT NULL,
+                        [Headers] [nvarchar](max) NULL,
+                        [Dimension] [nvarchar](50) NOT NULL,
```
Comment (sebastienros) [inline src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: I would suggest to go for a static set of columns instead, unless you feel confident with PBI.

File: src/Benchmarks/Startup.cs [hunk 11]
```diff
@@ -154,6 +154,12 @@ public void ConfigureServices(IServiceCollection services)
             {
                 services.AddResponseCaching();
             }
+
+            if (Scenarios.Any("SignalR"))
```
Comment (sebastienros) [inline src/Benchmarks/Startup.cs]: I am fine having them in Benchmarks, but you might prefer a simpler custom application to have more flexibility. Even in this repository.

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs (hunk 9 above)
Comment (BrennanConroy) [reply src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: Removed a couple items, ConnectionFilter and Method

File: src/BenchmarksWorkers/Workers/SignalRSerializer.cs (hunk 10 above)
Comment (BrennanConroy) [reply src/BenchmarksWorkers/Workers/SignalRSerializer.cs]: Seems fairly straightforward, we can talk offline

**Pull request merged**